File: desktopclient/src/main/scala/com/goexp/galgame/gui/task/TaskService.java
package com.goexp.galgame.gui.task;

import javafx.beans.value.WeakChangeListener;
import javafx.concurrent.Service;
import javafx.concurrent.Task;

import java.util.Objects;
import java.util.function.Supplier;


public class TaskService<V> extends Service<V> {

    private final Supplier<Task<V>> taskSupplier;

    public TaskService(Supplier<Task<V>> taskSupplier) {
        Objects.requireNonNull(taskSupplier);
        this.taskSupplier = taskSupplier;

        this.exceptionProperty().addListener(new WeakChangeListener<>((observable, oldValue, newValue) -> {
            if (newValue != null)
                newValue.printStackTrace();
        }));

    }

    @Override
    public Task<V> createTask() {
        return taskSupplier.get();
    }


}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/task/TaskService.scala
package com.goexp.galgame.gui.task

import java.util.Objects

import javafx.beans.value.WeakChangeListener
import javafx.concurrent.{Service, Task}

class TaskService[V](val taskCreator: () => Task[V]) extends Service[V] {
  Objects.requireNonNull(taskCreator)

  this.exceptionProperty.addListener {
    new WeakChangeListener[Throwable]((_, _, newValue) =>
      if (newValue != null)
        newValue.printStackTrace()
    )
  }

  override def createTask: Task[V] = taskCreator()
}

object TaskService {
  def apply[V](taskCreator: () => Task[V]) = new TaskService[V](taskCreator)
}